runtime/type.js

```diff
@@ -104,21 +109,35 @@ class Type {
       return Type.newInterface(shape);
     }
 
+    if (this.isConstraint) {
+      return Type.newConstraint(new Constraint(
+          this.constraint.variable.assignVariableIds(variableMap),
+          this.constraint.type.assignVariableIds(variableMap)));
+    }
+
     return this;
   }
 
   static unwrapPair(type1, type2) {
```

Comment:
May be worth unit test for this since it looks like it's added recursion and isn't intended to be a private method.

Reply:
Reverted.